File: backend/internal/api/history.go
package api

import (
	"fmt"
	"log"

	"github.com/gofiber/fiber/v2"
	"github.com/waluyo/wapbolt-backend/internal/middleware"
	"github.com/waluyo/wapbolt-backend/internal/repository"
)

type CreateHistoryRequest struct {
	TeamID          uint             `json:"team_id"`
	RequestID       uint             `json:"request_id"`
	Method          string           `json:"method"`
	URL             string           `json:"url"`
	RequestHeaders  repository.JSONB `json:"request_headers"`
	RequestBody     string           `json:"request_body"`
	ResponseHeaders repository.JSONB `json:"response_headers"`
	ResponseBody    string           `json:"response_body"`
	StatusCode      int              `json:"status_code"`
	ResponseTime    int              `json:"response_time"`
}

func SetupHistoryRoutes(app *fiber.App) {
	history := app.Group("/api/v1/history")
	history.Use(middleware.RequireAuth)

	history.Get("/", GetTeamHistory)
	history.Post("/", CreateHistory)
	history.Delete("/:id", DeleteHistory)
	history.Delete("/", ClearTeamHistory)
}

func GetTeamHistory(c *fiber.Ctx) error {
	teamID := c.Query("team_id")
	if teamID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": "team_id query param is required"})
	}

	if !canAccessTeam(c, parseUint(teamID)) {
		return c.Status(fiber.StatusForbidden).JSON(fiber.Map{"error": "Forbidden"})
	}

	var histories []repository.RequestHistory
	repository.DB.Preload("User").Where("team_id = ?", teamID).Order("created_at desc").Limit(50).Find(&histories)
	return c.JSON(histories)
}

func CreateHistory(c *fiber.Ctx) error {
	userId := c.Locals("user_id").(float64)
	var req CreateHistoryRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": "Invalid request body", "code": "BAD_REQUEST"})
	}

	if !canAccessTeam(c, req.TeamID) {
		return c.Status(fiber.StatusForbidden).JSON(fiber.Map{"error": "Forbidden"})
	}

	history := repository.RequestHistory{
		UserID:          uint(userId),
		TeamID:          req.TeamID,
		RequestID:       req.RequestID,
		Method:          req.Method,
		URL:             req.URL,
		RequestHeaders:  req.RequestHeaders,
		RequestBody:     req.RequestBody,
		ResponseHeaders: req.ResponseHeaders,
		ResponseBody:    req.ResponseBody,
		StatusCode:      req.StatusCode,
		ResponseTime:    req.ResponseTime,
	}

	if err := repository.DB.Create(&history).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": "Failed to save history", "code": "INTERNAL_SERVER_ERROR"})
	}

	return c.Status(fiber.StatusCreated).JSON(history)
}

func DeleteHistory(c *fiber.Ctx) error {
	id := c.Params("id")
	userId := c.Locals("user_id").(float64)

	if err := repository.DB.Where("id = ? AND user_id = ?", id, uint(userId)).Delete(&repository.RequestHistory{}).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": "Failed to delete history", "code": "INTERNAL_SERVER_ERROR"})
	}

	return c.JSON(fiber.Map{"message": "History deleted"})
}

func ClearTeamHistory(c *fiber.Ctx) error {
	teamID := c.Query("team_id")
	fmt.Println("sini teamID", teamID)
	if teamID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": "team_id query param is required"})
	}

	if !isAdminOrAbove(c, parseUint(teamID)) {
		log.Printf("Access denied to clear team history for team %s", teamID)
		return c.Status(fiber.StatusForbidden).JSON(fiber.Map{"error": "Only Admin or Owner can clear team history"})
	}

	if err := repository.DB.Where("team_id = ?", teamID).Delete(&repository.RequestHistory{}).Error; err != nil {
		fmt.Println("sini ", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"error": "Failed to clear history", "code": "INTERNAL_SERVER_ERROR"})
	}

	return c.JSON(fiber.Map{"message": "Team history cleared"})
}

File: apps/desktop-local/src/main/local/handlers/history.ts
import type Database from 'better-sqlite3'
import { Row, getString, getFloat, nowIso, LOCAL_USER_ID } from './helpers'
import { historyToJson } from './serializers'

// Port dari backend/internal/api/history.go. History local-only (§1 non-goal
// sync) — tidak menyentuh sync_meta, hard delete langsung.

type Res = { status: number; data: unknown }

export function getTeamHistory(db: Database.Database, query: URLSearchParams): Res {
  const teamId = query.get('team_id')
  if (!teamId) {
    return { status: 400, data: { error: 'team_id query param is required' } }
  }
  const rows = db
    .prepare('SELECT * FROM request_history WHERE team_id = ? ORDER BY created_at DESC LIMIT 50')
    .all(teamId) as Row[]
  return { status: 200, data: rows.map(historyToJson) }
}

export function createHistory(db: Database.Database, body: Row | null): Res {
  if (!body) {
    return { status: 400, data: { error: 'Invalid request body', code: 'BAD_REQUEST' } }
  }
  const result = db
    .prepare(
      `INSERT INTO request_history
        (user_id, team_id, request_id, method, url, request_headers, request_body,
         response_headers, response_body, status_code, response_time, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      LOCAL_USER_ID,
      getFloat(body, 'team_id'),
      getFloat(body, 'request_id'),
      getString(body, 'method'),
      getString(body, 'url'),
      JSON.stringify(body.request_headers ?? {}),
      getString(body, 'request_body'),
      JSON.stringify(body.response_headers ?? {}),
      getString(body, 'response_body'),
      getFloat(body, 'status_code'),
      getFloat(body, 'response_time'),
      nowIso()
    )
  const row = db
    .prepare('SELECT * FROM request_history WHERE id = ?')
    .get(Number(result.lastInsertRowid)) as Row
  return { status: 201, data: historyToJson(row) }
}

export function deleteHistory(db: Database.Database, id: string): Res {
  // Go: scoped ke user pemilik — lokal semua milik user 1.
  db.prepare('DELETE FROM request_history WHERE id = ? AND user_id = ?').run(id, LOCAL_USER_ID)
  return { status: 200, data: { message: 'History deleted' } }
}

export function clearTeamHistory(db: Database.Database, query: URLSearchParams): Res {
  const teamId = query.get('team_id')
  if (!teamId) {
    return { status: 400, data: { error: 'team_id query param is required' } }
  }
  db.prepare('DELETE FROM request_history WHERE team_id = ?').run(teamId)
  return { status: 200, data: { message: 'Team history cleared' } }
}
